packages/massa-web3/src/experimental/basicElements/address.ts

```diff
@@ -43,58 +45,102 @@ export class Address {
    *
    * @throws If the version is not supported.
    */
-  protected static initFromVersion(version: Version = Version.V0): Address {
+  protected static initFromVersion(
+    version: Version = DEFAULT_VERSION
+  ): Address {
     switch (version) {
       case Version.V0:
-        return new Address(new Base58(), new VarintVersioner(), version, false)
+        return new Address(new Base58(), new VarintVersioner(), version)
       default:
         throw new Error(`unsupported version: ${version}`)
     }
   }
 
+  private getPrefix(str: string): string {
+    const expected = [ADDRESS_USER_PREFIX, ADDRESS_CONTRACT_PREFIX]
+    for (let prefix of expected) {
+      if (str.startsWith(prefix)) {
+        return prefix
+      }
+    }
+    throw new Error(
+      `invalid address prefix: one of ${expected.join(' or ')} was expected.`
+    )
+  }
+
   /**
    * Initializes a new address object from a serialized string.
    *
    * @param str - The serialized address string.
-   * @param version - The version of the address. If not defined, the last version will be used.
    *
    * @returns A new address instance.
    *
    * @throws If the address string is invalid.
    */
-  public static fromString(str: string, version?: Version): Address {
+  public static fromString(str: string): Address {
+    const version = Address.getVersion(str)
     const address = Address.initFromVersion(version)
 
     try {
-      checkPrefix(
-        str,
-        ADDRESS_PREFIX + ADDRESS_USER_PREFIX,
-        ADDRESS_PREFIX + ADDRESS_CONTRACT_PREFIX
-      )
-      address.isEOA = str[ADDRESS_PREFIX.length] === ADDRESS_USER_PREFIX
-      address.bytes = extractData(
-        address.serializer,
-        address.versioner,
-        str.slice(ADDRESS_PREFIX.length + ADDRESS_USER_PREFIX.length),
-        address.version
+      const prefix = address.getPrefix(str)
+
+      address.isEOA = prefix === ADDRESS_USER_PREFIX
+      const versionedBytes = address.serializer.deserialize(
+        str.slice(prefix.length)
       )
+
+      address.bytes = Uint8Array.from([
+        ...varint.encode(address.isEOA ? 0 : 1),
+        ...versionedBytes,
+      ])
     } catch (e) {
       throw new Error(`invalid address string: ${e.message}`)
     }
-
     return address
   }
 
+  /**
+   * Get the address type from bytes.
+   *   *
+   * @returns the address type enum.
+   */
+  private getType(): number {
+    if (!this.bytes) {
+      throw new Error('address bytes is not initialized')
+    }
+    return varint.decode(this.bytes)
+  }
+
+  /**
+   * Get the address version.
+   *   *
+   * @returns the address type enum.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  static getVersion(data: string | Uint8Array): Version {
```

Comment:
I don't see the point of having this function public as it will add complexity. Except if you have an use case in mind.